rust/kernel/debugfs.rs

```diff
@@ -0,0 +1,201 @@
+// anything with a `&DebugFsDirectory`. This makes it is safe to share across
+// threads. [`DebugFsDirectory::create`] does take a `&DebugFsDirectory`, but it
+// is only publicly accessible through [`DebugFsDirectory::create_with_parent`]
+// which requires the reference to be mutable.
+unsafe impl Sync for DebugFsDirectory {}
+
+impl DebugFsDirectory {
+    /// Create a new directory in `debugfs` under `parent`. If `parent` is
+    /// `None`, it will be created at the `debugfs` root. The directory will be
+    /// recursively removed on drop.
+    fn create(name: &CStr, parent: Option<&DebugFsDirectory>) -> Result<Self> {
+        let name = name.as_char_ptr();
+        let parent_ptr = match parent {
+            Some(dir) => dir.dentry,
+            None => ptr::null_mut(),
+        };
+        // SAFETY: Calling a C function. `name` is a valid null-terminated
+        // string because it came from a [`CStr`] and `parent` is either null or
+        // valid because it came from a [`DebugFsDirectory`].
+        let dentry =
+            error::from_kernel_err_ptr(unsafe { bindings::debugfs_create_dir(name, parent_ptr) })?;
+        Ok(DebugFsDirectory {
+            dentry,
+            has_parent: parent.is_some(),
+            file_children: Vec::new(),
+            dir_children: Vec::new(),
+        })
+    }
+
+    /// Create a new directory in `debugfs` under `parent`. The directory will
+    /// be removed when the parent is dropped.
+    pub fn create_with_parent<'name, 'parent>(
+        name: &'name CStr,
+        parent: &'parent mut DebugFsDirectory,
+    ) -> Result<&'parent mut DebugFsDirectory> {
+        let result = DebugFsDirectory::create(name, Some(parent))?;
+        parent.dir_children.try_push(result)?;
+        let index = parent.dir_children.len() - 1;
+        Ok(&mut parent.dir_children[index])
+    }
+
+    /// Create a new directory at the toplevel of `debugfs`.
+    /// The directory will be recursively removed on drop.
+    pub fn create_toplevel(name: &CStr) -> Result<DebugFsDirectory> {
+        DebugFsDirectory::create(name, None)
+    }
+}
+
+impl Drop for DebugFsDirectory {
+    fn drop(&mut self) {
+        // If this entry has a parent, we don't need to worry about removal
+        // because the parent will remove its children when dropped. Otherwise
+        // we need to clean up.
+        if !self.has_parent {
+            // SAFETY: Calling a C function. `dentry` must have been created by
+            // a call to [`DebugFsDirectory::create`] which always returns a
+            // valid `dentry`. There is no parent, so the `dentry` couldn't have
+            // been removed and must still be valid.
+            unsafe {
+                bindings::debugfs_remove(self.dentry);
+            }
+        }
+    }
+}
+
+/// An `inode` for a file in debugfs with a `T` stored in `i_private`.
+pub struct DebugFsFile<T: PointerWrapper> {
+    dentry: *mut bindings::dentry,
+    data: *mut c_types::c_void,
```

Comment:
If the data is stored in the dentry's inode, why do we need to hold a pointer to it here?